src/Scriban/Syntax/ScriptVisitor.cs

```diff
@@ -0,0 +1,767 @@
+// Copyright (c) Alexandre Mutel. All rights reserved.
+// Licensed under the BSD-Clause 2 license.
+// See license.txt file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Scriban.Syntax
+{
+    public class ScriptVisitor
+    {
+        private readonly ScriptVisitorContext _context = new ScriptVisitorContext();
+
+        public IScriptVisitorContext Context => _context;
```

Comment:
Visitor pattern implementations don't typically have that kind of thing, but I found it useful for my use case to be able to access the ancestors of the current node. It could go away if `ScriptNode` had a `Parent` property.